### PR TITLE
An example using new TEMPLATE notation.

### PR DESCRIPTION
**Pull request by tclune**
The files in this directory are an attempt to demonstrate several
ideas that were bantered about at the last subgroup telecon using
the AXPY algorithm as an example.

  1. Introduce a new TEMPLATE construct - sort of like an inner
     parameterized module.
  2. How can a map procedure be combined with a rank-agnostic Array
     wrapper template, to provide operations on the wrapped elements?
  3. Use procedures instead of operators as template arguments.  (More
     general, and sidesteps some anoying notation issues to a degree.)

### CODE REVIEW

**Comment by tclune** [comment]
Note that this example also does not allow direct operations between "raw" intrinsics (scalar or array), but instead requires all operands to be wrapped with the generic Array wrapper.

File: scenarios/axpy_with_arrays/AXPY.F90 [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      REQUIRES assignable(T_y, T_axpy)
+
+   contains
+
+      subroutine AXPY(a, x, y)
+         type(T), intent(in) :: a
+         type(U), intent(in) :: x
+         type(V), intent(inout) :: y
+
+         y = plus(times(a,x), y)
```
Comment (rouson) [inline scenarios/axpy_with_arrays/AXPY.F90:49]: This is crying out for actual operator notation.   I have many times used lines like this to demonstrate the considerable improvement in clarity that operators provide.  The AXPY case isn't so bad, but one only has to add a few more terms to the expression for it to rapidly descend into being essentially indecipherable without translating it to operator syntax.

File: scenarios/axpy_with_arrays/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+2. We have no way to specify that arrays are conformable.  There is a
+   runtime aspect of this that is inescapable, as the underlying
+   shapes can be different, but in theory I would like the `map_t`
+   template to require that either all ranks are equal except those
+   that are scalar:
+	 - `(rank_U == rank_T) .or. rank_u*rank_T = 0`
+     - `rank_V == max(rank_T, rank_U)`
+   Do we want to be able to do this?
+3. Procedures might be more general, but this example really does
+   scream for using operators.
```
Comment (rouson) [inline scenarios/axpy_with_arrays/README.md:51]: Funny.  Sounds like we're on the same page.  My "crying out" is your "screaming."  LOL

File: scenarios/axpy_with_arrays/AXPY.F90 [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+   RESTRICTION assignable(T, U)
+      elemental assignment(=)(x,y)
+      type(T), intent(out) :: x 
+      type(U), intent(in) :: y
+      end assignment
+   END RESTRICTION
+
+   template axpy_t(T_a, T_x, T_y, T_ax, T_axpy, times, plus)
+
+      type :: T_a
```
Comment (rouson) [inline scenarios/axpy_with_arrays/AXPY.F90:24]: double colons are optional without an attribute so you might remove them here and in many places below

**Review by rouson — APPROVED** [review]

File: scenarios/axpy_with_arrays/Array.F90 [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      end type t
+      integer :: rank
+
+      type :: Array
+         type(T), rank(rank), allocatable :: elements
+      end type Array
+      
+      interface assignment(=)
+         module procedure assign_from_raw
+         module_procedure assign_to_raw
```
Comment (zoziha) [inline scenarios/axpy_with_arrays/Array.F90:16]:
```suggestion
         module procedure assign_to_raw
```
May be a minor typo here :)

File: scenarios/axpy_with_arrays/AXPY.F90 (hunk 1 above)
Comment (tclune) [reply scenarios/axpy_with_arrays/AXPY.F90:49]: Oh certainly.  But subgroup decided procedures were the priority.  I wanted to see how severe the consequences were ...   I really/unfortunately think we need both.

File: scenarios/axpy_with_arrays/AXPY.F90 (hunk 3 above)
Comment (tclune) [reply scenarios/axpy_with_arrays/AXPY.F90:24]: But ... that's the way I roll ... :-)

File: scenarios/axpy_with_arrays/Array.F90 (hunk 4 above)
Comment (tclune) [reply scenarios/axpy_with_arrays/Array.F90:16]:
```suggestion
         module procedure assign_to_raw
```
yup

**Pull request merged**